File: dot-net-CSharp/ExceptionHandlerExample.Api/Exceptions/Base/CustomizedException.cs
using Models.Response.Messages;

namespace Infra.CrossCutting.Exceptions
{
    public class CustomizedException : Exception
    {
        public ExceptionType ExceptionType {  get; }
        protected string tracking {  get; }

        protected CustomizedException(ExceptionType ExceptionType, string message) : base(message)
        {
           this.ExceptionType = ExceptionType;
           this.tracking = Guid.NewGuid().ToString("N");
        }

        protected CustomizedException(ExceptionType ExceptionType) : base()
        {
           this.ExceptionType = ExceptionType;
           this.tracking = Guid.NewGuid().ToString("N");
        }

        public static CustomizedException ofTimeOut(string message = null) {
            return of(ExceptionType.TIMEOUT, message);
        }

        public static CustomizedException ofBadGateway(string message = null) {
            return of(ExceptionType.BAD_GATEWAY, message);
        }

        public static CustomizedException ofServiceUnavailable(string message = null) {
            return of(ExceptionType.SERVICE_UNAVAILABLE, message);
        }

        public static CustomizedException ofInfraestructure(string message = null) {
            return of(ExceptionType.INFRASTRUCTURE, message);
        }

        public static CustomizedException of(string message = null) {
            return of(ExceptionType.EXCEPTION, message);
        }

        private static CustomizedException of(ExceptionType errorType, string message = null) {
            var errorDetails = message ?? ResponseMessages.Messages[errorType];
            return new CustomizedException(errorType, errorDetails);
        }
    }
}

File: java/exception-handler-example-api/src/main/java/com/example/infrastructure/Exceptions/base/CustomizedException.java
package com.example.infrastructure.Exceptions.base;

import java.util.Objects;

import com.example.infrastructure.Exceptions.enums.ExceptionType;

public class CustomizedException extends RuntimeException implements Trackable {

    protected ExceptionType exceptionType;
    protected String tracking;

    protected CustomizedException(ExceptionType exceptionType, String message) {
        super(message);
        this.tracking = java.util.UUID.randomUUID().toString();
    }

    public static CustomizedException ofTimeOut() {
        return CustomizedException.of(ExceptionType.TIMEOUT);
    }

    public static CustomizedException ofTimeOut(String message) {
        return CustomizedException.of(ExceptionType.TIMEOUT, message);
    }

    public static CustomizedException ofBadGateway() {
        return of(ExceptionType.BAD_GATEWAY);
    }

    public static CustomizedException ofBadGateway(String message) {
        return of(ExceptionType.BAD_GATEWAY, message);
    }

    public static CustomizedException ofServiceUnavailable() {
        return of(ExceptionType.SERVICE_UNAVAILABLE);
    }

    public static CustomizedException ofServiceUnavailable(String message) {
        return of(ExceptionType.SERVICE_UNAVAILABLE, message);
    }

    public static CustomizedException ofInfraestructure() {
        return of(ExceptionType.INFRASTRUCTURE);
    }

    public static CustomizedException ofInfraestructure(String message) {
        return of(ExceptionType.INFRASTRUCTURE, message);
    }

    public static CustomizedException of(String message) {
        return of(ExceptionType.EXCEPTION, message);
    }

    private static CustomizedException of(ExceptionType errorType) {
        return new CustomizedException(errorType, errorType.getMessage());
    }

    private static CustomizedException of(ExceptionType errorType, String message) {
        var errorDetails = !Objects.isNull(message) ? message : errorType.getMessage();
        return new CustomizedException(errorType, errorDetails);
    }

    @Override
    public String getTracking() {
        return tracking;
    }

    public ExceptionType getExceptionType() {
        return exceptionType;
    }
}
